### PR TITLE
Badges add

### PR DESCRIPTION
**Pull request by desmatter**
Cron for earliest practice/quiz/reflections and top ten/highest scores. 

Also added inspirer badge functionality to the cron that runs every 15 minutes.

Changed updateBadgeProgress in order to support having additions to the badges in userProgress, as old code would overwrite them.

### CODE REVIEW

File: bin/cron.js [hunk 1]
```diff
@@ -88,6 +334,9 @@ cron.schedule("*/15 * * * *", async () => {
       });
       badges.value.map((badge) => {
         badgeIdToPoints[badge._id] = parseInt(badge.Points);
+        if (badge._id === 32 && typeof badge._id.assignment_id !== "undefined") {
```
Comment (jnoowin) [inline bin/cron.js]: `badge._id.assignment_id` ?? Isn't badge._id a number?

File: bin/cron.js [hunk 2]
```diff
@@ -101,6 +350,7 @@ cron.schedule("*/15 * * * *", async () => {
             apply: 0,
             reflection: 0,
             daily: 0,
+            inspirer: 0,
```
Comment (jnoowin) [inline bin/cron.js]: I feel like this is scope creep of this function. Again, it's purely for updating the modules. Couldn't this be in the cron job that you made above?

File: bin/cron.js [hunk 3]
```diff
@@ -158,7 +408,8 @@ async function updateModuleProgress(
   try {
     let score = 0;
     const moduleProgress = {};
-    submissions.map((submission) => {
+    for (const submission of submissions) {
```
Comment (jnoowin) [inline bin/cron.js:316]: Do you know why there it's showing a diff of changes that were already merged? I made this change a while ago. Did you pull overhaul after you committed or something?

File: bin/cron.js [hunk 4]
```diff
@@ -38,6 +38,252 @@ cron.schedule("0 0 0 * * *", () => {
...
+
+      // Retrieve all necessary information at once
+      const [userSubmissions, modules, badges] = await Promise.allSettled([
+        userSubmissionsPromise(),
+        modulesPromise(),
+        badgesPromise(),
+      ]);
+
+      //Intitalizes dicts of scores for practice and quizzes
+      let highestPractice = {};
```
Comment (jnoowin) [inline bin/cron.js]: In general, use `const` first, then `let`, and never `var`. `const` declarations don't let you change what it references, so you should usually always use it for arrays and objects, cause you typically only change their attributes, not their references. Use `let` for primitive that change, otherwise const.

File: bin/cron.js [hunk 5]
```diff
@@ -38,6 +38,252 @@ cron.schedule("0 0 0 * * *", () => {
...
+            apply: 0,
+            reflection: 0,
+            daily: 0,
+            inspirer: 0,
+          },
+          userProgress = await db // Get current user's progress from MongoDB
+            .collection("user_progress")
+            .findOne({ user: user.user_id.toString() });
+        if (!userProgress) continue;
+        score += await updateModuleProgress(
```
Comment (jnoowin) [inline bin/cron.js]: Why `updateModuleProgress` here?

File: bin/cron.js (hunk 1 above)
Comment (desmatter) [reply bin/cron.js]: Yea you're right will change it

File: bin/cron.js (hunk 2 above)
Comment (desmatter) [reply bin/cron.js]: Sure yea can put it in the other

File: bin/cron.js (hunk 3 above)
Comment (desmatter) [reply bin/cron.js:316]: Yea I'm not sure either, I think I may have made this branch off an old commit a bit ago

File: bin/cron.js (hunk 5 above)
Comment (desmatter) [reply bin/cron.js]: I was having trouble figuring out how to set the score properly with how it is set up, came up with the easiest way is to set up a dict that maps the score accumulated from each user's module progress, and then add the badge scores below to that score. Lmk if you can think of an easier way though

File: bin/cron.js (hunk 4 above)
Comment (desmatter) [reply bin/cron.js]: Got it

**Review by desmatter — COMMENTED** [review]
looks good man tested it and it seems to run well
